src/slice.c

```diff
@@ -77,16 +78,18 @@ static SEXP list_slice(SEXP x, SEXP index) {
 
 #undef SLICE_BARRIER
 
-static SEXP rows_slice(SEXP x, SEXP index) {
+static SEXP df_slice(SEXP x, SEXP index) {
   R_len_t n = Rf_length(x);
   SEXP out = PROTECT(Rf_allocVector(VECSXP, n));
 
   SEXP nms = Rf_getAttrib(x, R_NamesSymbol);
   Rf_setAttrib(out, R_NamesSymbol, nms);
 
   for (R_len_t i = 0; i < n; ++i) {
-    SEXP sliced = vec_slice_impl(VECTOR_ELT(x, i), index, true);
+    SEXP elt = PROTECT(VECTOR_ELT(x, i));
```

Comment:
Why does this need protecting?

Reply:
In case it is generated by an altrep vector. Though I don't know anything about protection standards in that case.

Reply:
I think ALTREP will have to handle protection, otherwise it would break a lot of existing code. @jimhester can you confirm? (or at least speculate given that I don't think we have ALTREP lists yet)

Reply:
We don't have altrep lists, and I think the semantics would have to be the same as normal lists, so you shouldn't need an additional PROTECT here.